util::bound seems broken on Clang with older libstdc++

Debian 9.2 with Clang 3.8.1 and libstdc++ 6.3.0.
Building HPX master commit f5041a83ac0c01d8dc6c0539fbc1491e97ebee42 in C++14 mode, for Debug.
`cmake ../hpx -DCMAKE_BUILD_TYPE=Debug -DBOOST_ROOT=/opt/boost-1.65.1-cxx14 -DCMAKE_CXX_COMPILER=clang++ -DHPX_WITH_CXX14=ON -G Ninja -DHPX_WITH_EXAMPLES=OFF -DHPX_WITH_DEPRECATION_WARNINGS=OFF`

Full build log: https://gist.github.com/zao/02941613711a11ce609fe8d7034e5bcd

```
01:50 <@K-ballo> seems like libstdc++ pre 8.1 would be the cause
01:53 <@K-ballo> i'll have to look into it.. libstdc++ seems to be asking whether it can construct an util::bound from an std::tuple
01:53 <@K-ballo> the answer would definitively be no, but it's a hard error sfinae unfriendly one
01:54 <@K-ballo> one of those bugs with single element tuple constructors
01:54 <@K-ballo> should be able to reproduce with an older libstdc++, and add a workaround for it
02:43 <@K-ballo> doesn't look like something my recent changes would have caused
```

```
In file included from /tree/hpx/src/runtime_impl.cpp:8:
In file included from /tree/hpx/hpx/performance_counters/counters.hpp:11:
In file included from /tree/hpx/hpx/lcos/future.hpp:12:
In file included from /tree/hpx/hpx/lcos/detail/future_data.hpp:24:
/tree/hpx/hpx/util/bind.hpp:206:21: error: no viable conversion from 'const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > >' to 'void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &)'
                    std::forward<F_>(f),
                    ^~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/type_traits:970:24: note: in instantiation of function template specialization 'hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> >::bound<const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &, void>' requested here
             = decltype(::new _Tp(declval<_Arg>()))>
                              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/type_traits:971:24: note: in instantiation of default argument for '__test<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> >, const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &>' required here
      static true_type __test(int);
                       ^~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/type_traits:981:24: note: while substituting deduced template arguments into function template '__test' [with _Tp = hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> >, _Arg = const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &, $2 = (no value)]
      typedef decltype(__test<_Tp, _Arg>(0)) type;
                       ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/type_traits:144:14: note: in instantiation of template class 'std::__is_direct_constructible_impl<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> >, const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &>' requested here
    : public conditional<_B1::value, _B2, _B1>::type
             ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/type_traits:986:14: note: in instantiation of template class 'std::__and_<std::is_destructible<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > >, std::__is_direct_constructible_impl<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> >, const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &> >' requested here
    : public __and_<is_destructible<_Tp>,
             ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/type_traits:1068:14: note: (skipping 6 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
    : public conditional<is_reference<_Tp>::value,
             ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/tuple:631:21: note: in instantiation of function template specialization 'std::_TC<true, hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > >::_MoveConstructibleTuple<const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &>' requested here
                    _MoveConstructibleTuple<_UElements...>()
                    ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/tuple:636:19: note: while substituting prior template arguments into non-type template parameter [with _UElements = <const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &>]
        constexpr tuple(_UElements&&... __elements)
                  ^~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/tuple:554:11: note: while substituting deduced template arguments into function template 'tuple' [with _UElements = <const std::tuple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > > &>, $1 = (no value)]
    class tuple : public _Tuple_impl<0, _Elements...>
          ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/thread:138:8: note: in instantiation of template class 'std::_Bind_simple<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> > ()>' requested here
              std::__bind_simple(std::forward<_Callable>(__f),
              ^
/tree/hpx/src/runtime_impl.cpp:451:24: note: in instantiation of function template specialization 'std::thread::thread<hpx::util::detail::bound<void (hpx::runtime_impl::*)(std::mutex &, std::condition_variable &, bool &), hpx::runtime_impl *, std::reference_wrapper<std::mutex>, std::reference_wrapper<std::condition_variable>, std::reference_wrapper<bool> >>' requested here
        compat::thread t (util::bind(
                       ^
```

--- a/hpx/util/bind.hpp
+++ b/hpx/util/bind.hpp
@@ -199,7 +199,7 @@ namespace hpx { namespace util
 
             template <typename F_, typename ...Ts_, typename =
                 typename std::enable_if<
-                    !std::is_same<typename std::decay<F_>::type, bound>::value
+                    std::is_constructible<F, F_>::value
                 >::type>
             HPX_CONSTEXPR explicit bound(F_&& f, Ts_&&... vs)
               : base_type{

--- a/hpx/util/bind_back.hpp
+++ b/hpx/util/bind_back.hpp
@@ -109,7 +109,7 @@ namespace hpx { namespace util
 
             template <typename F_, typename ...Ts_, typename =
                 typename std::enable_if<
-                    !std::is_same<typename std::decay<F_>::type, bound_back>::value
+                    std::is_constructible<F, F_>::value
                 >::type>
             HPX_CONSTEXPR explicit bound_back(F_&& f, Ts_&&... vs)
               : base_type{

--- a/hpx/util/bind_front.hpp
+++ b/hpx/util/bind_front.hpp
@@ -109,7 +109,7 @@ namespace hpx { namespace util
 
             template <typename F_, typename ...Ts_, typename =
                 typename std::enable_if<
-                    !std::is_same<typename std::decay<F_>::type, bound_front>::value
+                    std::is_constructible<F, F_>::value
                 >::type>
             HPX_CONSTEXPR explicit bound_front(F_&& f, Ts_&&... vs)
               : base_type{

--- a/hpx/util/deferred_call.hpp
+++ b/hpx/util/deferred_call.hpp
@@ -109,7 +109,7 @@ namespace hpx { namespace util
 
             template <typename F_, typename ...Ts_, typename =
                 typename std::enable_if<
-                    !std::is_same<typename std::decay<F_>::type, deferred>::value
+                    std::is_constructible<F, F_>::value
                 >::type>
             explicit HPX_HOST_DEVICE deferred(F_&& f, Ts_&&... vs)
               : base_type{

--- a/hpx/util/one_shot.hpp
+++ b/hpx/util/one_shot.hpp
@@ -33,12 +33,12 @@ namespace hpx { namespace util
               : _called(false)
             {}
 
-            HPX_CONSTEXPR explicit one_shot_wrapper(F const& f)
-              : _f(f)
-              , _called(false)
-            {}
-            HPX_CONSTEXPR explicit one_shot_wrapper(F&& f)
-              : _f(std::move(f))
+            template <typename F_, typename =
+                typename std::enable_if<
+                    std::is_constructible<F, F_>::value
+                >::type>
+            HPX_CONSTEXPR explicit one_shot_wrapper(F_&& f)
+              : _f(std::forward<F_>(f))
               , _called(false)
             {}
 
@@ -60,11 +60,12 @@ namespace hpx { namespace util
             HPX_CONSTEXPR one_shot_wrapper()
             {}
 
-            HPX_CONSTEXPR explicit one_shot_wrapper(F const& f)
-              : _f(f)
-            {}
-            HPX_CONSTEXPR explicit one_shot_wrapper(F&& f)
-              : _f(std::move(f))
+            template <typename F_, typename =
+                typename std::enable_if<
+                    std::is_constructible<F, F_>::value
+                >::type>
+            HPX_CONSTEXPR explicit one_shot_wrapper(F_&& f)
+              : _f(std::forward<F_>(f))
             {}
 
             HPX_CONSTEXPR one_shot_wrapper(one_shot_wrapper&& other)

